### PR TITLE
Add az vm repair reset-nic command

### PR DESCRIPTION
**Pull request by swbae31**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### Related command
Added new command:
az vm repair reset-nic

Sample call:
az vm repair reset-nic -g <resource-group-name> -n <vm-name> --verbose

* Adding as preview mode until next month CLI core release rolls out and the new 'az network vnet subnet list-available-ips' is available. (Current version uses 'az network vnet list-available-ips' and it might not work at all cases)

For easy testing run these in CLI:
1)    Open Cloud shell or Azure CLI console
2)    Run:
az extension remove -n vm-repair
az extension add --source https://azurecomputeaidrepair.blob.core.windows.net/test/vm_repair-0.4.4-py2.py3-none-any.whl

az vm repair reset-nic -g resource-group-name -n vm-name --verbose

### General Guidelines

- [V] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [V] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- NOT a new extension

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
The precondition is to put your code inside this repository and upgrade the version in the pull request but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
vm repair

**Comment by swbae31** [comment]
Not sure why the Extensions index is failing. It is passing on local:
image

**Comment by zhoxing-ms** [comment]
@swbae31 Known issue, it has been solved by PR #5119

File: src/vm-repair/azext_vm_repair/_help.py [hunk 1]
```diff
@@ -77,3 +77,18 @@
           text: >
             az vm repair list-scripts --preview "https://github.com/haagha/repair-script-library/blob/master/map.json"
 """
+
+helps['vm repair reset-nic'] = """
+    type: command
+    short-summary: Resets the network interface stack on the VM guest OS. https://docs.microsoft.com/en-us/troubleshoot/azure/virtual-machines/reset-network-interface
```
Comment (zhoxing-ms) [inline src/vm-repair/azext_vm_repair/_help.py:83]: Please use the first person voice in the help message
```suggestion
    short-summary: Reset the network interface stack on the VM guest OS. https://docs.microsoft.com/en-us/troubleshoot/azure/virtual-machines/reset-network-interface
```

File: src/vm-repair/azext_vm_repair/_params.py [hunk 2]
```diff
@@ -48,3 +48,6 @@ def load_arguments(self, _):
 
     with self.argument_context('vm repair list-scripts') as c:
         c.argument('preview', help="URL of forked repair script library's map.json https://github.com/{user}/repair-script-library/blob/master/map.json")
+
+    with self.argument_context('vm repair reset-nic') as c:
+        c.argument('yes', help='Auto starts the VM skipping prompt if it is not running.')
```
Comment (zhoxing-ms) [inline src/vm-repair/azext_vm_repair/_params.py:53]: Could we use a parameter name with clearer meaning, such as `--skip-prompt`?

**Comment by zhoxing-ms** [comment]
Could you write the description of PR change into the history notes `HISTORY.rst`?

File: src/vm-repair/azext_vm_repair/_params.py (hunk 2 above)
Comment (swbae31) [reply src/vm-repair/azext_vm_repair/_params.py:53]: Actually, '--yes' used by multiple az commands as a skip prompt parameter. 

For example: 
https://docs.microsoft.com/en-us/cli/azure/vm?view=azure-cli-latest#az-vm-delete
az vm delete [--force-deletion]
             [--ids]
             [--name]
             [--no-wait]
             [--resource-group]
             [--yes]


AND 
https://docs.microsoft.com/en-us/cli/azure/deploymentmanager/rollout?view=azure-cli-latest
az deploymentmanager rollout restart --name
                                     --resource-group
                                     [--skip-succeeded]
                                     [--yes]

and many more commands. I was following consistency among commands. I think --yes is reasonable.

Comment (swbae31) [reply src/vm-repair/azext_vm_repair/_params.py]: But I edited the param description to be more readable.

**Review by haagha — COMMENTED** [review]
@swbae31 Adding my comments.

**Comment by swbae31** [comment]
> 

History updated.

File: src/vm-repair/azext_vm_repair/custom.py [hunk 3]
```diff
@@ -529,3 +530,113 @@ def list_scripts(cmd, preview=None):
...
+            raise SupportingResourceNotFoundError('Available IP address was not found within the VM subnet.')
+
+        # 3) Update private IP address to another in subnet. This will invoke and wait for a VM restart.
+        logger.info('Updating VM IP configuration. This might take a few minutes...\n')
+        # Update IP address
+        update_ip_command = 'az network nic ip-config update -g {g} --nic-name {nic} -n {config} --private-ip-address {ip} ' \
+                            .format(g=resource_group_name, nic=primary_nic_name, config=ipconfig_name, ip=swap_ip_address)
+        _call_az_command(update_ip_command)
+
+        # 4) Change things back. This will also invoke and wait for a VM restart.
```
[the remark this answers is not in the document]
Comment (swbae31) [reply src/vm-repair/azext_vm_repair/custom.py:592]: the vm state running is checked at the beginning, so this shouldn't be an issue

File: src/vm-repair/azext_vm_repair/custom.py [hunk 4]
```diff
@@ -529,3 +530,113 @@ def list_scripts(cmd, preview=None):
...
+            # Raise available IP not found
+            raise SupportingResourceNotFoundError('Available IP address was not found within the VM subnet.')
+
+        # 3) Update private IP address to another in subnet. This will invoke and wait for a VM restart.
+        logger.info('Updating VM IP configuration. This might take a few minutes...\n')
+        # Update IP address
+        update_ip_command = 'az network nic ip-config update -g {g} --nic-name {nic} -n {config} --private-ip-address {ip} ' \
+                            .format(g=resource_group_name, nic=primary_nic_name, config=ipconfig_name, ip=swap_ip_address)
+        _call_az_command(update_ip_command)
+
```
[the remark this answers is not in the document]
Comment (swbae31) [reply src/vm-repair/azext_vm_repair/custom.py:591]: call az command has its own exception path. It is caught by the AzCommandError except block.

File: src/vm-repair/azext_vm_repair/repair_utils.py [hunk 5]
```diff
@@ -318,13 +363,10 @@ def _check_linux_hyperV_gen(source_vm):
         fetch_hypervgen_command = 'az vm get-instance-view --ids {id} --query "[instanceView.hyperVGeneration]" -o json'.format(id=source_vm.id)
         hyperVGen_list = loads(_call_az_command(fetch_hypervgen_command))
         hyperVGen = hyperVGen_list[0]
-        if hyperVGen == 'V2':
-            return hyperVGen
-        else:
+        if hyperVGen != 'V2':
             hyperVGen = 'V1'
```
[the remark this answers is not in the document]
Comment (swbae31) [reply src/vm-repair/azext_vm_repair/repair_utils.py:367]: there is no change in logic. I just fixed the styling of the if statements to follow styling guidelines.

**Comment by swbae31** [comment]
@[zhoxing-ms](https://github.com/zhoxing-ms) can this be merged now?

**Pull request merged**